### PR TITLE
Allow user to set number of cores for R functions

### LINKED ISSUE #695: Number of cores specified for multithreading in R pixel clustering scripts can exceed actual number of available CPUs
*Please refer to our [FAQ](https://ark-analysis.readthedocs.io/en/latest/_rtd/faq.html) and look at our [known issues](https://github.com/angelolab/ark-analysis/issues) before opening a bug report.*

**Describe the bug**

On certain computers, setting `nCores <- parallel::detectCores() - 1` can cause the call to `parallel::makeCluster` to fail, especially if several cores are occupied at the time. The issue is that `parallel::detectCores()` only tells you how many CPUs are on the machine without telling the user how many are actually available.

**Expected behavior**

The value assigned to `nCores` should not exceed the number of actual CPUs available at the moment.

Unfortunately, R does not provide a way to tell us this directly. There are a couple ways around this:

- Default to `parallel::detectCores() - 1`, but allow the user to specify number of cores if needed
- Default to `parallel::detectCores() / 2`, not a fan of this option because it's overkill in most cases

### PR DESCRIPTION
**Pull request by alex-l-kong**
**What is the purpose of this PR?**

Closes #695. The way R parallel processing runs, sometimes the number of cores needs to be adjusted to prevent failures.

**How did you implement your changes**

Add a customizable argument `ncores` to the pixel SOM assignment and pixel consensus clustering functions, which get passed to their respective R scripts.

**Remaining issues**

Is this a function we want to add for pixel data preprocessing and pixel meta cluster renaming?

### CODE REVIEW

**Comment by ngreenwald** [comment]
The main issue was the lack of error messages, right? We need to make sure the user actually knows to adjust this parameter

**Comment by alex-l-kong** [comment]
@ngreenwald yeah, I'm working on that right now.

File: ark/phenotyping/pixel_consensus_cluster.R [hunk 1]
```diff
@@ -112,23 +128,24 @@ for (batchStart in seq(1, length(fovs), batchSize)) {
...
-        data.frame(fov=fovs[i], status=status)
+        data.frame(fov=fovs[i], status=status[1], errCond=status[2])
     }
 
     # report any erroneous feather files
     for (i in 1:nrow(fovStatuses)) {
         if (fovStatuses[i, 'status'] == 1) {
-            print(paste("The data for FOV", fovStatuses[i, 'fov'], "has been corrupted, removing"))
+            print(paste("Processing for FOV", fovStatuses[i, 'fov'], "failed, removing from pipeline. Error message:"))
+            print(fovStatuses[i, 'errCond'])
```
Comment (ngreenwald) [inline ark/phenotyping/pixel_consensus_cluster.R:148]: Were you able to find an easy way to test what happens when too many cores are specified, confirming that this now shows the correct error message?

File: ark/phenotyping/run_pixel_som.R [hunk 2]
```diff
@@ -97,23 +110,24 @@ for (batchStart in seq(1, length(fovs), batchSize)) {
...
-        data.frame(fov=fovs[i], status=status)
+        data.frame(fov=fovs[i], status=status[1], errCond=status[2])
     }
 
     # report any erroneous feather files
     for (i in 1:nrow(fovStatuses)) {
         if (fovStatuses[i, 'status'] == 1) {
-            print(paste("The data for FOV", fovStatuses[i, 'fov'], "has been corrupted, removing"))
+            print(paste("Processing for FOV", fovStatuses[i, 'fov'], "failed, removing from pipeline. Error message:"))
+            print(fovStatuses[i, 'errCond'])
```
Comment (ngreenwald) [inline ark/phenotyping/run_pixel_som.R:130]: Same here

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looks good, see below

File: ark/phenotyping/pixel_consensus_cluster.R (hunk 1 above)
Comment (alex-l-kong) [reply ark/phenotyping/pixel_consensus_cluster.R:148]: @ngreenwald yes, set the number of cores to an extremely high value (say, 128 on a laptop) and it now displays the error message `"Too many cores specified"`.

File: ark/phenotyping/run_pixel_som.R (hunk 2 above)
Comment (alex-l-kong) [reply ark/phenotyping/run_pixel_som.R:130]: @ngreenwald same as above.

**Review by ngreenwald — APPROVED** [review]
Great, looks good

**Pull request merged**